Side load snapshot

ffakenz opened the pull request:
<!-- Describe your change here -->

Closes https://github.com/cardano-scaling/hydra/issues/1858

## Summary
🛩️  introduce new SideLoadSnapshot ClientInput

🛩️  introduce new endpoint POST /snapshot
  - calls new ClientInput

🛩️  introduce new ServerOutput SnapshotSideLoaded
  - to signal when SideLoadSnapshot has been performed.

🛩️  HeadLogic now handles SideLoadSnapshot
  - persists LocalStateCleared event: if InitialSnapshot has been side loaded
  - persists LocalStateCleared event + SnapshotConfirmed: if ConfirmedSnapshot has been side loaded
  - produce SnapshotSideLoaded server output

🛩️  introduce new LogicError SideLoadSnapshotFailed
  - to signal when SideLoadSnapshot fails.

🛩️  HeadLogic handles event much like the SnapshotConfirmed
  - the new snapshot its been adopted as latest confirmed.

🛩️  HeadLogic handles event LocalStateCleared
  - It clears pending transactions in the local state, resets the local UTxO, and restarts the snapshot signing round by pruning the latest seen snapshot, which also resets the current leader.

🛩️  introduce new endpoint GET /snapshot
  - returns latest ConfirmedSnapshot

🛩️  introduce new endpoint GET /snapshot
  - returns latest ConfirmedSnapshot
  - useful in combination with new POST /snapshot in order to reset local state to latest known.

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
draft e2e scenario for head getting stuck

ffakenz committed:
add GET /snapshot

ffakenz committed:
add new client input for side load snapshot

ffakenz committed:
introduce server output, outcome event and aggregate handle in logic

ffakenz committed:
draft logic handle

ffakenz committed:
add POST /snapshot and POST /snapshot/latest

ffakenz committed:
fix head stuck

ffakenz committed:
Update CHANGELOG

ffakenz committed:
update docs

ffakenz committed:
fixes after rebase

v0d1ch commented on docs/docs/how-to/operating-hydra.md:
```diff
@@ -70,3 +70,19 @@ hydra_head_tx_confirmation_time_ms_count  0
+
+As a result of this divergence, the local ledger state of each Hydra node essentially becomes forked, resulting in inconsistent states across the network. While it is technically still possible to submit transactions to the Hydra nodes, doing so is ineffective because snapshots do not update unless all nodes sign them. Each node starts accepting transactions based on entirely different states, le ...
+
+To recover from this issue, we introduced side-loading of snapshots to synchronize the local ledger state of the Hydra nodes. With this mechanism, every peer reverts to the latest confirmed snapshot. This can be done using the POST /snapshot/latest endpoint, which clears the peer’s local pending transactions and restores its state to match the last agreed snapshot, allowing the node to rejoin the ...
```
In the changelog you mention `POST /snapshot` but here the route is different, here `POST /snapshot/latest endpoint`

v0d1ch replied:
I see... there are two routes actually.

v0d1ch commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -1305,6 +1308,112 @@ canDecommit tracer workDir node hydraScriptsTxId =
+        send n3 $ input "NewTx" ["transaction" .= tx]
+
+        -- Everyone confirms it
+        -- Note: We can't use `waitForAlMatch` here as it expects them to
```
```suggestion
        -- Note: We can't use `waitForAllMatch` here as it expects them to
```

v0d1ch commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -225,6 +227,61 @@ channels:
+      operationId: sideLoadLatestConfirmedSnapshotRequest
+      message:
+        payload:
+          type: "null"
```
Do we need `message` here if `payload` is null?

ffakenz replied:
good catch!
apparently we can!

v0d1ch commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -936,8 +936,7 @@ onOpenChainDepositTx headId env st deposited depositTxId deadline =
       <> if not snapshotInFlight && isLeader parameters party nextSn
-        then
-          cause (NetworkEffect $ ReqSn version nextSn (txId <$> localTxs) Nothing (Just deposited))
+        then cause (NetworkEffect $ ReqSn version nextSn (txId <$> localTxs) Nothing (Just deposited))
```
This line can be reverted

v0d1ch commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1248,6 +1333,14 @@ update env ledger st ev = case (st, ev) of
       | otherwise ->
           Error NotOurHead{ourHeadId, otherHeadId = headId}
+  (Open openState@OpenState{headId = ourHeadId}, ClientInput (SideLoadSnapshot confirmedSnapshot)) ->
+    let otherHeadId =
```
You could use `getSnapshot` function here and then just extract the `headId` from the `Snapshot`

v0d1ch commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1108,6 +1107,92 @@ onOpenChainCloseTx openState newChainState closedSnapshotNumber contestationDead
+  case sideLoadConfirmedSnapshot of
+    InitialSnapshot{} ->
+      if currentConfirmedSnapshot == sideLoadConfirmedSnapshot
+        then -- Spec:  ̅S ← snObj(v̂, ŝ, Û, T̂, 𝑈𝛼, 𝑈𝜔)
```
Do these additions require spec changes? If so are they already done? I am not sure if you should have spec comments here unless this piece of code has relevant spec parts.

ffakenz replied:
Agree.
I added them just for reference but I agree this feature is not part of the normal user journey thus shouldn't be included in our specs.

v0d1ch commented on hydra-node/src/Hydra/Logging/Monitoring.hs:
```diff
@@ -94,6 +94,7 @@ monitor transactionsMap metricsMap = \case
     -- transactions after some timeout expires
     atomically $ modifyTVar' transactionsMap (Map.insert (txId tx) t)
     tick "hydra_head_requested_tx"
+  -- REIVEW! should we handle SnapshotSideLoaded ???
```
I think it would be a nice addition since it is as important as `SnapshotConfirmed` imo

ch1bo replied:
We discussed this on friday: We **must** have only one event that can be counted as a `SnapshotConfirmed`.

Additional info like a side load happened can go into another `StateEvent` / `ServerOutput`. That means, the head logic which does a side load should result in two state changes.

v0d1ch commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -722,6 +724,117 @@ spec =
+        let tx1 = SimpleTx 1 mempty (utxoRef 2) -- No inputs, requires no specific starting state
+            tx2 = SimpleTx 2 (utxoRef 2) (utxoRef 3)
+            tx3 = SimpleTx 3 (utxoRef 3) (utxoRef 4)
+            snapshot1 = Snapshot testHeadId 0 1 [tx1] (utxoRef 2) Nothing Nothing
```
you could test also how  `utxoToDe/Commit` is behaving - related to your review comment in the head logic.

v0d1ch reviewed (approved):
Nice work @ffakenz Perhaps we should also have a mob review session since if feels like it would be beneficial to go through more possible scenarios together.

ffakenz committed:
comment rewording

Co-authored-by: Sasha Bogicevic <Sasha.Bogicevic@iohk.io>

ffakenz committed:
remove message payload from api spec

* as not needed

ffakenz committed:
remove spec details on side load snapshot client input

* as it is not part of normal user journey, thus, it is not part of our formal spec.

ffakenz committed:
make use of getSnapshot when checking input has expected head id

ffakenz committed:
add minor review comment

* also removing unnintended macOS file

ffakenz committed:
minor revert

ffakenz committed:
minor fmt

ffakenz committed:
apply fixes from #1892 on with prepared hydra-node

ffakenz committed:
minor fmt

ffakenz commented on docs/docs/how-to/operating-hydra.md:
```diff
@@ -70,3 +70,19 @@ hydra_head_tx_confirmation_time_ms_count  0
+
+As a result of this divergence, the local ledger state of each Hydra node essentially becomes forked, resulting in inconsistent states across the network. While it is technically still possible to submit transactions to the Hydra nodes, doing so is ineffective because snapshots do not update unless all nodes sign them. Each node starts accepting transactions based on entirely different states, le ...
+
+To recover from this issue, we introduced side-loading of snapshots to synchronize the local ledger state of the Hydra nodes. With this mechanism, every peer reverts to the latest confirmed snapshot. This can be done using the POST /snapshot/latest endpoint, which clears the peer’s local pending transactions and restores its state to match the last agreed snapshot, allowing the node to rejoin the ...
```
remove POST /snapshot/latest

pipe GET + POST instead

ch1bo commented on hydra-node/src/Hydra/Logging/Monitoring.hs:
```diff
@@ -96,6 +96,7 @@ monitor transactionsMap metricsMap = \case
   (Node LogicOutcome{outcome = Continue{stateChanges}}) -> do
     forM_ stateChanges $ \case
+      -- REIVEW! should we handle SnapshotSideLoaded ???
       SnapshotConfirmed{snapshot = Snapshot{confirmed}} -> do
```
Must: have only one event that can be counted as a SnapshotConfirmed.

Additional info like a side load happened can go into another `StateEvent` / `ServerOutput`. That means, the head logic which does a side load should result in two state changes.

ffakenz committed:
fix http server spec after rebase

ffakenz committed:
add required verification to be same as L1 snapshot

To avoid accepting snapshots which can not be use to close the head with.

Also refactor solution:
* split SnapshotSideLoaded into SnapshotConfirmed + ClearLocalState
** reuses SnapshotConfirmed outcome and server outcome
** simplifies SnapshotSideLoaded to be produced upon ClearLocalState

ffakenz committed:
enhance client input and logic handle haddocks

ffakenz committed:
update CHANGELOG

ffakenz committed:
remove POST /snapshot/latest endpoint

* as trivial

ffakenz committed:
remove from api spec CommitIgnored after rebase

ffakenz committed:
fix e2e spec

* missing to reset localUTxO upon ClearLocalState event

ch1bo commented on hydra-node/src/Hydra/HeadLogic/Outcome.hs:
```diff
@@ -143,6 +143,7 @@ data StateChanged tx
       , participants :: [OnChainId]
       }
   | TxInvalid {headId :: HeadId, utxo :: UTxOType tx, transaction :: tx, validationError :: ValidationError}
+  | ClearLocalState {headId :: HeadId, snapshotNumber :: SnapshotNumber}
```
I like the minimalism of this `StateChanged` value and that it contains the `SnapshotNumber` to reference a `SnapshotConfirmed` event, but is this the best name for it? Sounds a lot like a "command" and how does a snapshot number relate to "clear local state"?

Could: find a better name

ffakenz replied:
how about: `LocalStateResetToSnapshot` ?

ffakenz replied:
going with `LocalStateCleared` because is hard to relate to the snapshot number  when in the future.

ch1bo commented on hydra-node/src/Hydra/Logging/Monitoring.hs:
```diff
@@ -96,6 +96,7 @@ monitor transactionsMap metricsMap = \case
     tick "hydra_head_requested_tx"
   (Node LogicOutcome{outcome = Continue{stateChanges}}) -> do
     forM_ stateChanges $ \case
+      -- REIVEW! should we handle SnapshotSideLoaded ???
```
Must: remove this outdated comment

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1127,6 +1128,90 @@ onOpenChainCloseTx openState newChainState closedSnapshotNumber contestationDead
+-- * something it was before (in the case of the initial snapshot).
+-- * something it would be using side communication (in the case of a confirmed snapshot).
+--
+-- Besides the above, it is expected to work very much like the confirmed snapshot.
```
Nice, good documentation!

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1127,6 +1128,90 @@ onOpenChainCloseTx openState newChainState closedSnapshotNumber contestationDead
+      | requestedSv /= lastSeenSv = Error $ RequireFailed $ ReqSvNumberInvalid{requestedSv, lastSeenSv}
+      | requestedSc /= lastSeenSc = Error $ AssertionFailed "ConfirmedSnapshot utxoToCommit side loaded does not match last known."
+      | requestedSd /= lastSeenSd = Error $ AssertionFailed "ConfirmedSnapshot utxoToDecommit side loaded does not match last known."
+      | otherwise = cont
```
Could: improve indentation here

I needed to read this function 2-3 times as it seemed at first that `cont` is an expression, but in fact it's an argument to the function.

```suggestion
  requireVerifiedL1Snapshot cont
      | requestedSv /= lastSeenSv = Error $ RequireFailed $ ReqSvNumberInvalid{requestedSv, lastSeenSv}
      | requestedSc /= lastSeenSc = Error $ AssertionFailed "ConfirmedSnapshot utxoToCommit side loaded does not match last known."
      | requestedSd /= lastSeenSd = Error $ AssertionFailed "ConfirmedSnapshot utxoToDecommit side loaded does not match last known."
      | otherwise = cont
```

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1127,6 +1128,90 @@ onOpenChainCloseTx openState newChainState closedSnapshotNumber contestationDead
+  requireVerifiedL1Snapshot
+    cont
+      | requestedSv /= lastSeenSv = Error $ RequireFailed $ ReqSvNumberInvalid{requestedSv, lastSeenSv}
+      | requestedSc /= lastSeenSc = Error $ AssertionFailed "ConfirmedSnapshot utxoToCommit side loaded does not match last known."
```
What is the difference between `RequireFailed` and `AssertionFailed`?

ffakenz replied:
I used AssertionFailed to avoid creating fine-grained RequirementFailures for each case.
wdyt?

ch1bo replied:
Is there a difference? If not, I'd suggest to add these to `RequireFailed`

Why? It's always a bad idea to have `data Error = Error Text` in some for another because it makes it "too easy" to add error cases and makes people (including our future selves) think less about what errors actually are there. If the barrier is non-zero, there is a chance for a developer to ask themselves.. can we avoid this error situation by-design? e.g. by pushing a problem upstream or make an invalid state unrepresentable.. That's where we want to end up: with less error cases.

ffakenz replied:
Defined its own error hierarchy to avoid conflicts with errors during ReqSn.

This also helps detect client-side errors during submission and respond accordingly.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1267,6 +1352,11 @@ update env ledger st ev = case (st, ev) of
+    let Snapshot{headId = otherHeadId} = getSnapshot confirmedSnapshot
+     in if ourHeadId == otherHeadId
+          then onOpenClientSideLoadSnapshot openState confirmedSnapshot
+          else Error NotOurHead{ourHeadId, otherHeadId}
```
This error will not reach the client. Neither do the `Error` outcomes in the `onOpenClientSideLoadSnapshot`.

Should: send responses to the client when something is wrong with the submitted snapshot.

ffakenz replied:
This is a great idea, but I can't find an easy way to achieve it.

> I thought about passing a callback from withAPIServer to runHydraNode, but the challenge is tying the callback to the request session.

I suggest we either pair on this or resolve it as a follow-up task, as it also affects other endpoints.

ch1bo replied:
We agreed on tuesday that this would be a follow-up task. @ffakenz please create one and link it here before resolving.

ffakenz replied:
https://github.com/cardano-scaling/hydra/issues/1911

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -1305,6 +1308,112 @@ canDecommit tracer workDir node hydraScriptsTxId =
+  withHydraNode hydraTracer aliceChainConfig workDir 1 aliceSk [bobVk, carolVk] [1, 2, 3] $ \n1 -> do
+    aliceUTxO <- seedFromFaucet cardanoNode aliceCardanoVk 1_000_000 (contramap FromFaucet tracer)
+    withHydraNode hydraTracer bobChainConfig workDir 2 bobSk [aliceVk, carolVk] [1, 2, 3] $ \n2 -> do
+      -- Carol starts its node missconfigured
```
```suggestion
      -- Carol starts its node misconfigured
```

ch1bo reviewed (approved):
Looks good now! Only some cleanup needed (see "Must:")

Not having any client feedback if the snapshot is not valid is not ideal, but we could also address this in a follow-up.

ffakenz committed:
fix typo on e2e scenario

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ffakenz committed:
refactor check on L1 Snapshot

* to enhance readability

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ffakenz committed:
remove outdated comment

ffakenz committed:
minor fmt

ffakenz committed:
fix: use requestedSn as snapshotNumber instead of lastSeenSn

ffakenz committed:
rename ClearLocalState to LocalStateCleared

* as it ounds a lot like a 'command'

ffakenz committed:
add new SideLoadSnapshotFailed logic error

ffakenz committed:
minor
